From 27bce69ea0caa953255d881d1dcd95af2d44b680 Mon Sep 17 00:00:00 2001
From: Meticulus <theonejohnnyd@gmail.com>
Date: Thu, 30 Nov 2017 04:17:04 -0600
Subject: [PATCH] "0001-met_dev_functions-meticulus
 _______________________________

*Previous From bbb338481f69b6d718771ac5fe11c09d24cb1581 Mon Sep 17 00:00:00 2001
*Previous From: Meticulus <theonejohnnyd@gmail.com>
*Previous Date: Thu, 30 Nov 2017 04:04:08 -0600
*Previous Subject: [PATCH] "0001-met_dev_functions-meticulus
*_______________________________
*
**Previous From 91134de21e803b6d7596cb08b89ade8e59baa4cc Mon Sep 17 00:00:00 2001
**Previous From: Meticulus <theonejohnnyd@gmail.com>
**Previous Date: Wed, 29 Nov 2017 08:16:08 -0600
**Previous Subject: [PATCH] "0001-met_dev_functions-meticulus
**_______________________________
**
***Previous From e32c1b55e798a56efcfd29e78087302be20cf8d8 Mon Sep 17 00:00:00 2001
***Previous From: Meticulus <theonejohnnyd@gmail.com>
***Previous Date: Wed, 29 Nov 2017 08:07:16 -0600
***Previous Subject: [PATCH] "0001-met_dev_functions-meticulus
***_______________________________
***
****Previous From 6487e35fdbc641d0ce7eb2edcda1e0dbc28e69e9 Mon Sep 17 00:00:00 2001
****Previous From: Meticulus <theonejohnnyd@gmail.com>
****Previous Date: Wed, 29 Nov 2017 08:05:12 -0600
****Previous Subject: [PATCH] "0001-met_dev_functions-meticulus
****_______________________________
****
*****Previous From 9b191ed9cb226c37a484b7952c8fe4c9b3116561 Mon Sep 17 00:00:00 2001
*****Previous From: Meticulus <theonejohnnyd@gmail.com>
*****Previous Date: Wed, 29 Nov 2017 07:55:36 -0600
*****Previous Subject: [PATCH] "0001-met_dev_functions-meticulus
*****_______________________________
*****
******Previous From aabde1617d766c43218fd0c80aa318d7914b7c59 Mon Sep 17 00:00:00 2001
******Previous From: Meticulus <theonejohnnyd@gmail.com>
******Previous Date: Sat, 25 Nov 2017 08:43:43 -0600
******Previous Subject: [PATCH] "0001-met_dev_functions-meticulus
******_______________________________
******
*******Previous From a65a60dd1d5fdc4e354c9e41e3d964fc4c914321 Mon Sep 17 00:00:00 2001
*******Previous From: Meticulus <theonejohnnyd@gmail.com>
*******Previous Date: Sat, 25 Nov 2017 08:33:56 -0600
*******Previous Subject: [PATCH] "0001-met_dev_functions-meticulus
*******_______________________________
*******
********Previous From 7c7b43934c886876a717dded25700ed648299959 Mon Sep 17 00:00:00 2001
********Previous From: Meticulus <theonejohnnyd@gmail.com>
********Previous Date: Thu, 23 Nov 2017 13:36:04 -0600
********Previous Subject: [PATCH] "0001-met_dev_functions-meticulus
********_______________________________
********
*********Previous From 9083c3c43177db87ea7b5e3968f050211daf626d Mon Sep 17 00:00:00 2001
*********Previous From: Meticulus <theonejohnnyd@gmail.com>
*********Previous Date: Mon, 19 Jun 2017 13:30:28 -0500
*********Previous Subject: [PATCH] "0001-met_dev_functions-meticulus
*********_______________________________
*********
**********Previous From a306a9e4c3cefb54ee13f814bffd370c1871b6ee Mon Sep 17 00:00:00 2001
**********Previous From: Meticulus <theonejohnnyd@gmail.com>
**********Previous Date: Mon, 19 Jun 2017 11:27:36 -0500
**********Previous Subject: [PATCH] "0001-met_dev_functions-meticulus
**********_______________________________
**********
***********Previous From bcee8fcd31aad8d79218bc507475be50ad302494 Mon Sep 17 00:00:00 2001
***********Previous From: Meticulus <theonejohnnyd@gmail.com>
***********Previous Date: Mon, 19 Jun 2017 11:20:22 -0500
***********Previous Subject: [PATCH] "0001-met_dev_functions-meticulus
***********_______________________________
***********
************Previous From b830fa6d02997e06a7208839561145ce2df1bd30 Mon Sep 17 00:00:00 2001
************Previous From: Meticulus <theonejohnnyd@gmail.com>
************Previous Date: Mon, 19 Jun 2017 11:13:28 -0500
************Previous Subject: [PATCH] "0001-met_dev_functions-meticulus
************_______________________________
************
*************Previous From 54db4b632bb650d7762d4982c897e3a4c4cc8e1a Mon Sep 17 00:00:00 2001
*************Previous From: Meticulus <theonejohnnyd@gmail.com>
*************Previous Date: Mon, 19 Jun 2017 11:03:03 -0500
*************Previous Subject: [PATCH] "0001-met_dev_functions-meticulus
*************_______________________________
*************
**************Previous From 7a6cc7c3307bb8b28a601f42b8cefaabf9f37041 Mon Sep 17 00:00:00 2001
**************Previous From: Meticulus <theonejohnnyd@gmail.com>
**************Previous Date: Mon, 19 Jun 2017 10:57:30 -0500
**************Previous Subject: [PATCH] "0001-met_dev_functions-meticulus
**************_______________________________
**************
***************Previous From 8fe2283c5a8113f6668da8efae98d1f88efe5ba0 Mon Sep 17 00:00:00 2001
***************Previous From: Meticulus <theonejohnnyd@gmail.com>
***************Previous Date: Mon, 19 Jun 2017 10:56:57 -0500
***************Previous Subject: [PATCH] "0001-met_dev_functions-meticulus
***************_______________________________
***************
****************Previous From 27f19f7902a9f4877dafbd2a7398e91d0977c97a Mon Sep 17 00:00:00 2001
****************Previous From: Meticulus <theonejohnnyd@gmail.com>
****************Previous Date: Mon, 19 Jun 2017 10:54:38 -0500
****************Previous Subject: [PATCH] "0001-met_dev_functions-meticulus
****************_______________________________
****************
*****************Previous From e06b44b6a7ce8b46200c2f5157b53643e2d282c4 Mon Sep 17 00:00:00 2001
*****************Previous From: Meticulus <theonejohnnyd@gmail.com>
*****************Previous Date: Sun, 18 Jun 2017 12:44:22 -0500
*****************Previous Subject: [PATCH] "0001-met_dev_functions-meticulus
*****************_______________________________
*****************
******************Previous From 0f3e72b354dd416926a90555fa719cf7d3ba6a0a Mon Sep 17 00:00:00 2001
******************Previous From: Meticulus <theonejohnnyd@gmail.com>
******************Previous Date: Tue, 1 Nov 2016 08:20:06 -0500
******************Previous Subject: [PATCH] 0001-met_dev_functions
******************"
*****************
*****************Change-Id: Ie77c1103bfcdac98b92c53766aa93f6641454eb1"
****************
****************Change-Id: I64a9142f219b193ecc05af21d0a59596cbd81440"
***************
***************Change-Id: I57a737c80824dbfebb0092439cd9f1ed9c8afc0f"
**************
**************Change-Id: I57bef2b57e69df0cfebf1b88d132a0115b02746f"
*************
*************Change-Id: Idab349f1cc9241e0714916718202f17ea4660121"
************
************Change-Id: Ied9c80af12d85cce85308b2ada608234183ddc5a"
***********
***********Change-Id: Icd2b4a0b87b671f883cea81f16561b05f41898f3"
**********
**********Change-Id: Idb4a7a01c5a68f7b4f74a804f177b7b03b6af8da"
*********
*********Change-Id: I5b35cc9fe2ee9751d3d01ee6b952b6ebfb328abb"
********
********Change-Id: Ib335621a1c1806e5e562e8f5e88b7c720295e2a8"
*******
*******Change-Id: I66fb69de04021962f7909411f39cf16390cf4788"
******
******Change-Id: I213880f1ba987b1aa8885c4f44da57477bcb5741"
*****
*****Change-Id: If51fb100d3d9fd9cadb1eb9580a66b9f5697ce64"
****
****Change-Id: I5d500c16870fd7329f16036c354c710bf8bf195c"
***
***Change-Id: Id5249f98f4f2e17c2245574b32c4c183909f9a1d"
**
**Change-Id: I8f3819ad9c4c6afc48f47aa2f79190b7afbce8a7"
*
*Change-Id: I582cfa95de7ed4272fa5d0535d842ea06f7bfe8b"

Change-Id: I1b0adc0d63a8cb9570210a751fac58f7355468f9
---
 gui/action.cpp                |  73 +++++++++++++++++++++++
 gui/objects.hpp               |   6 ++
 gui/resources.cpp             |   2 +-
 gui/theme/common/portrait.xml | 134 ++++++++++++++++++++++++++++++++++++++++++
 4 files changed, 214 insertions(+), 1 deletion(-)

diff --git a/gui/action.cpp b/gui/action.cpp
index a395158..bf80257 100644
--- a/gui/action.cpp
+++ b/gui/action.cpp
@@ -231,6 +231,12 @@ GUIAction::GUIAction(xml_node<>* node)
 		ADD_ACTION(twcmd);
 		ADD_ACTION(setbootslot);
 		ADD_ACTION(installapp);
+		ADD_ACTION(flashmagisk);
+		ADD_ACTION(flashusnf);
+		ADD_ACTION(usbhoston);
+		ADD_ACTION(usbhostoff);
+		ADD_ACTION(dump);
+		ADD_ACTION(restoredump);
 	}
 
 	// First, get the action
@@ -1760,6 +1766,73 @@ int GUIAction::twcmd(std::string arg)
 	return 0;
 }
 
+int GUIAction::flashmagisk(std::string arg)
+{
+	int wipe_cache = 0;
+	int op_status = 0;
+	operation_start("Flash Magisk");
+	op_status = flash_zip("/sbin/magisk.zip", &wipe_cache);
+	operation_end(op_status);
+	return 0;
+}
+
+int GUIAction::flashusnf(std::string arg)
+{
+	int wipe_cache = 0;
+	int op_status = 0;
+	operation_start("Flash Universal SafetyNet Fix");
+	op_status = flash_zip("/sbin/usnf.zip", &wipe_cache);
+	operation_end(op_status);
+	return 0;
+}
+
+int GUIAction::usbhoston(std::string arg)
+{
+	int op_status = 0;
+	operation_start("USB HOST ON");
+	op_status = system("echo hoston > /sys/devices/platform/ff100000.hisi_usb/plugusb");
+	if(!op_status) {
+		system(" echo 255 > /sys/class/leds/blue/brightness");
+		gui_msg("USB host mode on");
+	} else 
+		gui_warn("Could not turn on usb host!?!");
+	operation_end(op_status);
+	return 0;
+}
+
+int GUIAction::usbhostoff(std::string arg)
+{
+	int op_status = 0;
+	operation_start("USB HOST OFF");
+	op_status = system("echo hostoff > /sys/devices/platform/ff100000.hisi_usb/plugusb");
+	if(!op_status) {
+		system(" echo 0 > /sys/class/leds/blue/brightness");
+		gui_msg("USB host mode off");
+	} else 
+		gui_warn("Could not turn off usb host!?!");
+	operation_end(op_status);
+	return 0;
+}
+int GUIAction::dump(std::string arg)
+{
+	int wipe_cache = 0;
+	int op_status = 0;
+	operation_start("Dump");
+	op_status = flash_zip("/sbin/StockBackup_hi6250.zip", &wipe_cache);
+	operation_end(op_status);
+	return 0;
+}
+
+int GUIAction::restoredump(std::string arg)
+{
+	int wipe_cache = 0;
+	int op_status = 0;
+	operation_start("Restore Dump");
+	op_status = flash_zip("/sbin/RestoreStockBackup_hi6250.zip", &wipe_cache);
+	operation_end(op_status);
+	return 0;
+}
+
 int GUIAction::getKeyByName(std::string key)
 {
 	if (key == "home")		return KEY_HOMEPAGE;  // note: KEY_HOME is cursor movement (like KEY_END)
diff --git a/gui/objects.hpp b/gui/objects.hpp
index 630cf71..736916c 100644
--- a/gui/objects.hpp
+++ b/gui/objects.hpp
@@ -367,6 +367,12 @@ protected:
 	int twcmd(std::string arg);
 	int setbootslot(std::string arg);
 	int installapp(std::string arg);
+	int flashmagisk(std::string arg);
+	int flashusnf(std::string arg);
+	int usbhoston(std::string arg);
+	int usbhostoff(std::string arg);
+	int dump(std::string arg);
+	int restoredump(std::string arg);
 
 	int simulate;
 };
diff --git a/gui/resources.cpp b/gui/resources.cpp
index bd71d50..f07fe72 100644
--- a/gui/resources.cpp
+++ b/gui/resources.cpp
@@ -312,7 +312,7 @@ std::string ResourceManager::FindString(const std::string& name) const
 		std::map<std::string, string_resource_struct>::const_iterator it = mStrings.find(name);
 		if (it != mStrings.end())
 			return it->second.value;
-		LOGERR("String resource '%s' not found. No default value.\n", name.c_str());
+		//LOGERR("String resource '%s' not found. No default value.\n", name.c_str());
 		PageManager::AddStringResource("NO DEFAULT", name, "[" + name + ("]"));
 	} else {
 		LOGINFO("String resources not loaded when looking for '%s'. No default value.\n", name.c_str());
diff --git a/gui/theme/common/portrait.xml b/gui/theme/common/portrait.xml
index f163bea..e3d72cc 100644
--- a/gui/theme/common/portrait.xml
+++ b/gui/theme/common/portrait.xml
@@ -3514,6 +3514,12 @@
 					<action function="page">confirm_action</action>
 				</actions>
 			</button>
+                        
+			<button style="main_button">	
+				<placement x="%center_x%" y="%row19_y%"/>
+				<text>Extras</text>
+				<action function="page">meticulus</action>
+			</button>
 
 			<action>
 				<touch key="home"/>
@@ -3526,6 +3532,134 @@
 			</action>
 		</page>
 
+		<page name="meticulus">
+			<template name="page"/>
+
+			<text style="text_l">
+				<placement x="%col1_x_header%" y="%row3_header_y%"/>
+				<text>{@advanced_hdr=Advanced}</text>
+			</text>
+
+			<text style="text_m">
+				<placement x="%col1_x_header%" y="%row4_header_y%"/>
+				<text>Meticulus Development: Extras</text>
+			</text>
+
+			<button style="main_button">
+				<placement x="%indent%" y="%row2a_y%"/>
+				<text>Magisk</text>
+				<actions>
+					<action function="set">tw_back=meticulus</action>
+					<action function="set">tw_action=flashmagisk</action>
+					<action function="set">tw_has_action2=0</action>
+					<action function="set">tw_action_text1=Flashing Magisk</action>
+					<action function="set">tw_action_text2=</action>
+				        <action function="set">tw_text1=Flash Magisk</action>
+                                	<action function="set">tw_text2=This is intended for Nougat ROMs</action>
+                                	<action function="set">tw_text3=and may not work in Oreo+ ROMs</action>
+                                	<action function="set">tw_text4=</action>
+					<action function="set">tw_complete_text1=Magisk Flashed!</action>
+					<action function="set">tw_slider_text=Flash Magisk?</action>
+					<action function="page">confirm_action</action>	
+				</actions>
+				<action function="flash"/>
+			</button>
+			<button style="main_button">
+				<placement x="%center_x%" y="%row2a_y%"/>
+				<text>SafetyNet Fix</text>
+				<actions>
+					<action function="set">tw_back=meticulus</action>
+					<action function="set">tw_action=flashusnf</action>
+					<action function="set">tw_has_action2=0</action>
+					<action function="set">tw_action_text1=Flashing Universal SafetyNet Fix</action>
+					<action function="set">tw_action_text2=</action>
+				        <action function="set">tw_text1=Flash Universal Saftey Net Fix</action>
+                                	<action function="set">tw_text2=This is a Magisk Module and Magisk</action>
+                                	<action function="set">tw_text3=should be flashed first.</action>
+                                	<action function="set">tw_text4=</action>
+					<action function="set">tw_complete_text1=Universal SafetyNet Fix Flashed!</action>
+					<action function="set">tw_slider_text=Flash Universal SafteyNet Fix?</action>
+					<action function="page">confirm_action</action>	
+				</actions>
+				<action function="flash"/>
+			</button>
+
+                        <!--<button style="main_button">
+                                <placement x="%indent%" y="%row8_y%"/>
+                                <text>USB Host: ON</text>
+                                <actions>
+                                        <action function="set">tw_back=meticulus</action>
+                                        <action function="set">tw_action=usbhoston</action>
+                                        <action function="set">tw_has_action2=0</action>
+                                        <action function="set">tw_action_text1=Activating USB HOST</action>
+                                        <action function="set">tw_action_text2=</action>
+                                        <action function="set">tw_complete_text1=USB HOST ON</action>
+                                        <action function="page">action_page</action>
+                                </actions>
+                        </button>
+
+                        <button style="main_button">
+                                <placement x="%center_x%" y="%row8_y%"/>
+                                <text>USB Host: OFF</text>
+                                <actions>
+                                        <action function="set">tw_back=meticulus</action>
+                                        <action function="set">tw_action=usbhostoff</action>
+                                        <action function="set">tw_has_action2=0</action>
+                                        <action function="set">tw_action_text1=Activating USB HOST</action>
+                                        <action function="set">tw_action_text2=</action>
+                                        <action function="set">tw_complete_text1=USB HOST OFF</action>
+                                        <action function="page">action_page</action>
+                                </actions>
+                        </button>-->
+
+			<button style="main_button">
+				<placement x="%indent%" y="%row13a_y%"/>
+				<text>Dump</text>
+				<actions>
+					<action function="set">tw_back=meticulus</action>
+					<action function="set">tw_action=dump</action>
+					<action function="set">tw_has_action2=0</action>
+					<action function="set">tw_action_text1=Dumping system and boot</action>
+					<action function="set">tw_action_text2=</action>
+				        <action function="set">tw_text1=Dump system and boot</action>
+                                	<action function="set">tw_text2=This will create a dump of the system and boot partitions.</action>
+                                	<action function="set">tw_text3=in the /sdcard/meticulus folder. Use "Restore Dump" to</action>
+                                	<action function="set">tw_text4=return to stock.</action>
+					<action function="set">tw_complete_text1=Dump complete!</action>
+					<action function="set">tw_slider_text=Dump system and boot?</action>
+					<action function="page">confirm_action</action>	
+				</actions>
+			</button>
+			<button style="main_button">
+				<placement x="%center_x%" y="%row13a_y%"/>
+				<text>Restore Dump</text>
+				<actions>
+					<action function="set">tw_back=meticulus</action>
+					<action function="set">tw_action=restoredump</action>
+					<action function="set">tw_has_action2=0</action>
+					<action function="set">tw_action_text1=Restoring system and boot</action>
+					<action function="set">tw_action_text2=</action>
+				        <action function="set">tw_text1=Restore system and boot</action>
+                                	<action function="set">tw_text2=This will restore the dumps of system and boot created</action>
+                                	<action function="set">tw_text3=using "Dump", and format /data so that you can</action>
+                                	<action function="set">tw_text4=return to stock.</action>
+					<action function="set">tw_complete_text1=Restore dump complete</action>
+					<action function="set">tw_slider_text=Restore dump?</action>
+					<action function="page">confirm_action</action>	
+				</actions>
+			</button>
+
+			<action>
+				<touch key="home"/>
+				<action function="page">main</action>
+			</action>
+
+			<action>
+				<touch key="back"/>
+				<action function="page">advanced</action>
+			</action>
+		</page>
+
 		<page name="partsdcardsel">
 			<template name="page"/>
 
-- 
2.7.4

